x-pack/plugins/apm/public/components/app/transaction_details/waterfall_with_summary/waterfall_container/waterfall/waterfall_helpers/waterfall_helpers.ts

```diff
@@ -415,6 +416,20 @@ function getErrorCountByParentId(
   }, {});
 }
 
+export const getHasOrphanTraceItems = (
+  traceDocs: Array<WaterfallTransaction | WaterfallSpan>
+) => {
+  const waterfallItemsIds = traceDocs.map((doc) =>
+    doc.processor.event === 'span'
+      ? (doc?.span as WaterfallSpan['span']).id
+      : doc?.transaction?.id
+  );
+
+  return traceDocs.some(
+    (item) => item.parent?.id && !waterfallItemsIds.includes(item.parent?.id)
+  );
```

Comment:
The waterfall itself is already very slow, WDYT of this change to speed this function up? With this you change it from `O(n)` to `O(1)`.

```suggestion
  const waterfallItemsIds = new Set(traceDocs.map((doc) =>
    doc.processor.event === 'span'
      ? (doc?.span as WaterfallSpan['span']).id
      : doc?.transaction?.id
  ));

  return traceDocs.some(
    (item) => item.parent?.id && !waterfallItemsIds.has(item.parent.id)
  );
```

FYI: I did not test it.

Reply:
Nice idea, thank you! I changed that and tested it, looks good 👍